name-registry/src/main.sw

```diff
@@ -0,0 +1,165 @@
+contract;
+
+dep data_structures;
+dep errors;
+dep events;
+dep interface;
+
+use data_structures::Record;
+use errors::Errors;
+use events::{IdentityChanged, NameRegistered, OwnerChanged, RegistrationExtended};
+use interface::NameRegistry;
+use std::{
+    block::timestamp,
+    chain::auth::msg_sender,
+    constants::{
+        BASE_ASSET_ID,
+    },
+    context::call_frames::msg_asset_id,
+    context::msg_amount,
+    logging::log,
+    option::Option,
+    result::Result,
+    revert::{
+        require,
+        revert,
+    },
+    storage::StorageMap,
+};
+
+storage {
+    /// A mapping of names to an option of records, with a none representing an unregistered name
+    names: StorageMap<str[8], Option<Record>> = StorageMap {},
+}
+
+/// The amount to charge per hundred seconds per name
+const PRICE_PER_HUNDRED: u64 = 1;
+
+impl NameRegistry for Contract {
+    #[storage(read)]
+    fn expiry(name: str[8]) -> u64 {
+        match storage.names.get(name) {
+            Option::Some(record) => {
+                record.expiry
+            },
+            None => {
+                log(Errors::NameNotRegistered);
+                revert(0)
+            }
+        }
+    }
+
+    #[storage(read, write)]
+    fn extend(name: str[8], duration: u64) {
+        require(storage.names.get(name).is_some(), Errors::NameNotRegistered);
+        require(duration / 100 * PRICE_PER_HUNDRED <= msg_amount(), Errors::InsufficientPayment);
+        require(msg_asset_id() == BASE_ASSET_ID, Errors::WrongAssetSent);
+
+        let old_record = storage.names.get(name).unwrap();
+        let new_record = Record {
+            expiry: old_record.expiry + duration,
+            identity: old_record.identity,
+            owner: old_record.owner,
+        };
+
+        storage.names.insert(name, Option::Some(new_record));
+
+        log(RegistrationExtended {
+            duration,
+            name,
+            new_expiry: new_record.expiry,
+        });
+    }
+
+    #[storage(read)]
+    fn identity(name: str[8]) -> Identity {
```

Comment:
I think this should check if the identity is expired and revert if it is. Otherwise the registration expires but still resolves to the same `Identity` even after it's expired, essentially enabling a lifelong registration.

Reply:
Well my thought process was that the expiry is really just a way for someone else to grab the name if they want it, if its expired and yet no one wants it then let the old owner keep it. But I can just change it to not have it be like that